Core/enable cxx17

roigcarlo opened the pull request:
**DO NOT MERGE**

Draft of changes for C++17. Wip. 
This PR is to address some problems observed in #9916 mostly regarding deprecation of features towards C++20.

List of current changes:
- All `unitary_function` and `binary_function` instances have been removed from the code.
- The `return_type` from the functions above has been replaced with `decltype()` when necessary
- Current instances of code that already used `decltype()` using an object have been replaced with the a `std::declval` of that object to avoid creating an instance:
    ```C++
    decltype(TDataType)
    ```

    Changes to:

    ```C++
    decltype(std::declval<TDataType>())
    ```

    This is specially important because of Dofs in `model_part.h`:

    ```C++
    typedef PointerVectorSet<DofType,
            SetIdentityFunction<DofType>,
            std::less<SetIdentityFunction<DofType>>::result_type,
            std::equal_to<SetIdentityFunction<DofType>>::result_type,
            DofType* > DofsArrayType;
    ```

    Changes to:

    ```C++
    typedef PointerVectorSet<DofType,
            SetIdentityFunction<DofType>,
            std::less<decltype(std::declval<IndexedObject>()(std::declval<DofType>()))>,
            std::equal_to<decltype(std::declval<IndexedObject>()(std::declval<DofType>()))>,
            DofType* > DofsArrayType;
    ```

    which has undermined some inconsistencies with `DofType` in several core classes (For instance `scheme.h`):

    ```C++
    /// DoF iterator type definition
    typedef typename PointerVectorSet<TDofType, IndexedObject>::iterator DofIterator;
    /// DoF constant iterator type definition
    typedef typename PointerVectorSet<TDofType, IndexedObject>::const_iterator DofConstantIterator;
    ```

    Should be defined as:

    ```C++
    /// DoF iterator type definition
    typedef typename PointerVectorSet<TDofType, 
            std::less<decltype(std::declval<IndexedObject>()(std::declval<DofType>()))>,
            std::equal_to<decltype(std::declval<IndexedObject>()(std::declval<DofType>()))>,
            DofType* >::iterator DofIterator;
    /// DoF constant iterator type definition
    typedef typename PointerVectorSet<TDofType, 
            std::less<decltype(std::declval<IndexedObject>()(std::declval<DofType>()))>,
            std::equal_to<decltype(std::declval<IndexedObject>()(std::declval<DofType>()))>,
            DofType* >::const_iterator DofConstantIterator;
    ```

    Otherwise it forces the `PointerVectorSet` of some classes to work with `DofType::Pointer` instead of `DofType*` Which could potentially have an impact in performance.
    Also there seems to be an inconsistency on whether to use `IndexedObject` or `SetIdentityFunction` for the Key/Hash of the `PointerVectorSet<DofType>`. I asume `IndexedObject` is the correct approach since otherwise the sorting function fails ( seems that only RomApplication was sorting dofs...)

roigcarlo committed:
Set standard to 17

roigcarlo committed:
Removing function auxiliars

roigcarlo committed:
Adapting container interfaces

roigcarlo committed:
Correcting usage of containers in code

roigcarlo committed:
Correcting tests

roigcarlo committed:
Incorrect dof_updater

roigcarlo committed:
Incorrect dof_updater again

roigcarlo committed:
Incorrect dof_updater again x2

roigcarlo committed:
Using IndexedObject has key/hash for Dof PointerVectorSets

roigcarlo committed:
Using IndexedObject has key/hash in the scheme

roigcarlo committed:
Changing solver to use Id's instead of objects

This change is done because now all instance of PointerVectorSet<Dof> use IndexedObject as hasher. If we want to retain the same behaviour we can change it so functions use IndexedObject->Id() as default TGetKey, but if no implications in performance I would say this is clearer.

roigcarlo committed:
Fixing tests

roigcarlo committed:
Missing tests

roigcarlo committed:
Use dof id

roigcarlo committed:
Setting 17

roigcarlo committed:
More C17

roigcarlo committed:
Taht should not happen but this will protect the code more

roigcarlo committed:
Split std::optional implementation

roigcarlo committed:
Some more dof access by id

roigcarlo committed:
Fixing types

roigcarlo committed:
Restoring keys and fixing types

roigcarlo committed:
Reverted changes to dofs by Id

roigcarlo committed:
Deleted another binary function

roigcarlo committed:
Removing std::iterator inheritances

roigcarlo committed:
Trying to add traits to Project Parameters

roigcarlo committed:
Missing typenames

loumalouomega committed:
Merge branch 'master' into core/enable-cxx17

loumalouomega committed:
Revert "@roigcarlo, unable to fix, I hardcore the distance"

This reverts commit 1a1b18cf1f663eb4ada171991bb4085fede77595.

loumalouomega committed:
@roigcarlo, unable to fix, I hardcore the distance

roigcarlo committed:
Fixing centos flag

matekelemen committed:
make aliases public

matekelemen committed:
make aliases consistent with return types

matekelemen committed:
private member variables

loumalouomega committed:
Merge branch 'master' into core/enable-cxx17

philbucher commented on applications/CoSimulationApplication/custom_external_libraries/CoSimIO/CMakeLists.txt:
```diff
@@ -85,7 +85,7 @@ if(MSVC)
 
 elseif(${CMAKE_COMPILER_IS_GNUCXX})
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17")
```
this is a small standalone lib, this should not be changed I would say

matekelemen replied:
does that mean that ```CoSimIO``` will use a different ```libstdc++``` than the rest of Kratos? Is that safe?

roigcarlo replied:
Yes it is safe, its the same scenario as linking with a fotran module

philbucher replied:
Same as for any other system lib no?
Isnt that sorta garantied through ABI stability?

roigcarlo replied:
Well C++11 broke the ABI compatibility for the strings with old standards (remember the infamous `_GLIBCXX_USE_CXX11_ABI`). It is expectable, but not 100% guaranteed

matekelemen replied:
To be honest, I don't know much about compatibility and maintenance. I just noticed that all my binaries are updated whenever I get a new GCC version. And new C++ standards sometimes remove features / modify some signatures.

roigcarlo committed:
Revert back CoSimIO to C++11

roigcarlo committed:
Merge branch 'master' into core/enable-cxx17

roigcarlo committed:
Using new compiler in centos

roigcarlo committed:
Merge branch 'core/enable-cxx17' of https://github.com/KratosMultiphysics/Kratos into core/enable-cxx17

roigcarlo committed:
Enable gcc8 correctly

roigcarlo committed:
2nd Try

roigcarlo committed:
Merge remote-tracking branch 'origin/master' into core/enable-cxx17

roigcarlo committed:
Merge branch 'master' into core/enable-cxx17

roigcarlo committed:
Trying forcing if constexpr

roigcarlo committed:
Enabling back constrexp if

roigcarlo committed:
Disable 1011

source: https://community.intel.com/t5/Intel-C-Compiler/quot-if-constexpr-quot-and-quot-missing-return-statement-quot-in/td-p/1154551

philbucher committed:
refactor utility

philbucher committed:
using new fct

philbucher committed:
updated test

roigcarlo committed:
Fixing race condition

roigcarlo committed:
using correct communicator

roigcarlo committed:
Merge branch 'core/enable-cxx17' of https://github.com/KratosMultiphysics/Kratos into core/enable-cxx17

roigcarlo committed:
More legacy interface

philbucher committed:
adding SetMPICommunicatorRecursively

philbucher committed:
restructure process

philbucher committed:
using proper function

roigcarlo committed:
Merge branch 'master' into core/enable-cxx17

roigcarlo committed:
Merge branch 'master' into core/enable-cxx17

philbucher committed:
Merge branch 'master' into core/enable-cxx17

philbucher committed:
cleanup

philbucher committed:
disable intel deprecation warnings

philbucher committed:
cleanup

philbucher committed:
moving flag to CI

philbucher committed:
at least GCC 8 is required

philbucher committed:
let CMake handle the C++ standard

philbucher reviewed (approved):
I pushed some minor CI improvement, from my side it is ready :tada:
